src/Pirouette/Transformations/Monomorphization.hs

```diff
@@ -169,31 +191,54 @@ executeSpecRequest SpecRequest {origDef = HofDef {..}, ..} = M.fromList $
                ]
   where
     fixName = genSpecName specArgs
+    specArgsLen = length specArgs
+
+    -- When specializing constructor types, we need to substitute occurences of
+    -- the un-specialized type with the fixed name. For instance, if we're specializing
+    -- a constructor:
+    --
+    -- > Bin : all a : Type . Bin a -> Bin a -> Bin a
+    --
+    -- that was applied to TyBool, the result has to be:
+    --
+    -- > Bin@TyBool : Bin@TyBool -> Bin@TyBool -> Bin@TyBool
+    --
+    fixType = rewrite $ \case
+      SystF.TyApp (SystF.Free (TySig n)) xs -> do
+        guard (n == hofDefName && specArgs `isPrefixOf` xs)
+        return $ SystF.Free (TySig $ fixName n) `SystF.TyApp` drop specArgsLen xs
+      _ -> Nothing
 
 -- | Specializes a function application of the form:
 --
---  > hof @Integer x y z
+--  > hof @Integer @Bool x y z
 --
 -- at call site, where @hof@ has been identified as
 -- 1. either a higher-order function itself,
 -- 2. or invoking a polymorphic higher-order function, perhaps, transitively;
 -- and its type argument contains no bound-variables: in other words,
--- we can substitute that call with @hof_Integer@ while creating a monomorphic
+-- we can substitute that call with @hof\@Integer\@Bool@ while creating a monomorphic
 -- definition for @hof@.
 --
+-- We only specialize type arguments that appear /before/ the first term-argument.
+--
 -- This function only does the substitution _at call site_ and emits a 'SpecRequest' denoting that the corresponding
 -- higher-order _definition_ needs to be specialized (which will be handled later by 'executeSpecRequest').
 specFunApp :: forall lang. (LanguageBuiltins lang) => HOFDefs lang -> SpecFunApp lang
 specFunApp hofDefs (SystF.App (SystF.Free (TermSig name)) args)
-  | Just someDef <- name `M.lookup` hofDefs, -- We compare the entire name, not just the nameString part: x0 /= x1.
-    all isSpecArg $ take hofPolyVarsCount tyArgs = do
+  -- We compare the entire name, not just the nameString part: x0 /= x1.
+  | Just someDef <- name `M.lookup` hofDefs,
+    -- Now we ensure that there is something to specialize and that the type arguments we've
+    -- gathered are specializable arguments (ie, no bound type-variables)
+    hofPolyVarsCount >= 1,
```

Comment:
This is a very nitpicky thing, but I'm not a fan of patterns which relate to `where` (because I get confused by the scope). May I suggest changing it to something on the lines of the following?

```haskell
    -- with null
    not (null tyArgs),

    -- with pattern
    _ : _ <- tyArgs
```

If you then below inline `length tyArgs` in `splitArgs`, I think everything becomes more readable.